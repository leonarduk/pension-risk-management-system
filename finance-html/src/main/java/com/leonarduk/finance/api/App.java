package com.leonarduk.finance.api;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

import com.leonarduk.finance.stockfeed.Instrument.InstrumentLoader;

@SpringBootApplication
public class App {
	public static void main(final String[] args) throws Exception {
		InstrumentLoader.getInstance().init("resources/data/instruments_list.csv");
		
		SpringApplication.run(App.class, args);
	}
	

}
