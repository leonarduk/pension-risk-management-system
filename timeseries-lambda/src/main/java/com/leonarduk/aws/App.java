package com.leonarduk.aws;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Lists;
import com.leonarduk.finance.stockfeed.Instrument;
import com.leonarduk.finance.stockfeed.StockFeed;
import com.leonarduk.finance.stockfeed.feed.Commentable;
import com.leonarduk.finance.stockfeed.feed.yahoofinance.StockV1;
import com.leonarduk.finance.utils.DataField;
import com.leonarduk.finance.utils.HtmlTools;
import org.apache.commons.lang3.StringUtils;
import org.ta4j.core.Bar;

import java.io.IOException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Lambda function entry point. You can change to use other pojo type or implement
 * a different RequestHandler.
 *
 * @see <a href=https://docs.aws.amazon.com/lambda/latest/dg/java-handler.html>Lambda Java Handler</a> for more information
 */
public class App
        implements RequestHandler<APIGatewayProxyRequestEvent, APIGatewayProxyResponseEvent> {
    private final StockFeed stockFeed;

    public App() {
        stockFeed = DependencyFactory.stockFeed();
    }

    public static void main(String[] args) throws IOException {
        System.out.println(new App().getResults(ImmutableMap.of(
                "ticker", "PHGP.L",
                "years", "1",
                "interpolate", "true",
                "cleanData", "true"
        )));
    }

    @Override
    public APIGatewayProxyResponseEvent handleRequest(final APIGatewayProxyRequestEvent input, final Context context) {
        // TODO: invoking the api call using s3Client.
        APIGatewayProxyResponseEvent responseEvent = new APIGatewayProxyResponseEvent();
        try {
            responseEvent.setBody(getResults(input.getQueryStringParameters()));
            responseEvent.setStatusCode(200);
        } catch (IOException e) {
            e.printStackTrace();
            responseEvent.setBody("FAILED: " + e.getMessage());
            responseEvent.setStatusCode(500);
        }
        return responseEvent;
    }

    private String getResults(Map<String, String> inputParams) throws IOException {

        if (inputParams == null)
        {
            throw new IllegalArgumentException("No parameters provided. Expect at least ticker");
        }
        System.out.println(inputParams);


        String ticker = inputParams.get("ticker");
        final int years = Integer.parseInt(StringUtils.defaultIfEmpty(inputParams.get("years"), "10"));
        final int months = Integer.parseInt(StringUtils.defaultIfEmpty(inputParams.get("months"), "0"));
        final int weeks = Integer.parseInt(StringUtils.defaultIfEmpty(inputParams.get("weeks"), "0"));
        final int days = Integer.parseInt(StringUtils.defaultIfEmpty(inputParams.get("days"), "0"));

        final String fromDate = inputParams.get("fromDate");
        final String toDate = inputParams.get("toDate");
        final boolean interpolate = Boolean.parseBoolean(StringUtils.defaultIfEmpty(inputParams.get("interpolate"), "False"));
        final boolean cleanData = Boolean.parseBoolean(StringUtils.defaultIfEmpty(inputParams.get("cleanData"), "False"));

        String region = StringUtils.defaultIfEmpty(inputParams.get("region"), "L");
        String type = StringUtils.defaultIfEmpty(inputParams.get("type"), "UNKNOWN");
        String currency = StringUtils.defaultIfEmpty(inputParams.get("type"), "GBP");

        if (ticker.contains(".")) {
            String[] parts = ticker.split("\\.");
            ticker = parts[0];
            region = parts[1];
        }

        if (ticker.contains("/")) {
            String[] parts = ticker.split("/");
            ticker = parts[0];
            region = parts[1];

            if (parts.length > 2) {
                type = parts[2];
            }
            if (parts.length > 3) {
                currency = parts[3];
            }
        }
        final Instrument instrument = Instrument.fromString(ticker, region, type, currency);

        LocalDate toLocalDate;
        final LocalDate fromLocalDate;

        if (!StringUtils.isEmpty(fromDate)) {
            fromLocalDate = LocalDate.parse(fromDate);
            if (StringUtils.isEmpty(toDate)) {
                toLocalDate = LocalDate.now();
            } else {
                toLocalDate = LocalDate.parse(toDate);
            }
        } else {
            toLocalDate = LocalDate.now();
            if (days > 0) {
                fromLocalDate = LocalDate.now().plusDays(-1 * days);
            } else if (weeks > 0) {
                fromLocalDate = LocalDate.now().plusWeeks(-1 * weeks);
            } else if (months > 0) {
                fromLocalDate = LocalDate.now().plusMonths(-1 * months);
            } else {
                fromLocalDate = LocalDate.now().plusYears(-1 * years);
            }
        }


        return generateResults(fromLocalDate, toLocalDate, interpolate, cleanData, instrument);
    }

    private String generateResults(final LocalDate fromLocalDate, final LocalDate toLocalDate,
                                   final boolean interpolate, final boolean cleanData,
                                   final Instrument instrument)
            throws IOException {
        final StringBuilder sbBody = new StringBuilder();
        final List<List<DataField>> records = Lists.newArrayList();

        final List<Bar> historyData;

        historyData = this.getHistoryData(instrument, fromLocalDate, toLocalDate, interpolate, cleanData, false);

        for (final Bar historicalQuote : historyData) {
            final ArrayList<DataField> record = Lists.newArrayList();
            records.add(record);
            record.add(new DataField("Date", historicalQuote.getEndTime().toLocalDate().toString()));
            record.add(new DataField("Open", historicalQuote.getOpenPrice()));
            record.add(new DataField("High", historicalQuote.getMaxPrice()));
            record.add(new DataField("Low", historicalQuote.getMinPrice()));
            record.add(new DataField("Close", historicalQuote.getClosePrice()));
            record.add(new DataField("Volume", historicalQuote.getVolume()));

            if (historicalQuote instanceof Commentable) {
                Commentable commentable = (Commentable) historicalQuote;
                record.add(new DataField("Comment", commentable.getComment()));

            }
        }

        HtmlTools.printTable(sbBody, records);
        return HtmlTools.createHtmlText(null, sbBody).toString();
    }

    private List<Bar> getHistoryData(Instrument instrument, LocalDate fromLocalDate, LocalDate toLocalDate,
                                     boolean interpolate, boolean cleanData, boolean addLatestQuoteToTheSeries) throws IOException {
        final Optional<StockV1> stock = this.stockFeed.get(instrument, fromLocalDate, toLocalDate, interpolate,
                cleanData, addLatestQuoteToTheSeries);
        if (stock.isPresent()) {
            return stock.get().getHistory();
        }
        return Lists.newArrayList();
    }

}
