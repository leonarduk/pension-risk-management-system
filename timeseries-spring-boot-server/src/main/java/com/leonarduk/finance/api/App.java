package com.leonarduk.finance.api;

import com.leonarduk.finance.stockfeed.Instrument;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.core.env.Environment;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.web.bind.annotation.RestController;

import java.net.NetworkInterface;

@SpringBootApplication
@RestController
@EnableJpaRepositories(basePackageClasses = {InstrumentRepository.class})
@EntityScan(basePackageClasses=Instrument.class)
@ComponentScan("com.leonarduk...")
public class App {

    public static void main(final String[] args) throws Exception {
		SpringApplication.run(App.class, args);
	}


}
