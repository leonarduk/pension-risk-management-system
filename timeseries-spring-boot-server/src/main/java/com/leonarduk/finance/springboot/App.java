package com.leonarduk.finance.springboot;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Import;
import org.springframework.scheduling.annotation.EnableScheduling;

@SpringBootApplication
@EnableScheduling
@ComponentScan(basePackages = {"com.leonarduk.finance.springboot"})
@Import(AppConfig.class)
public class App {
	public static void main(final String[] args) throws Exception {
		SpringApplication.run(App.class, args);
	}


}
