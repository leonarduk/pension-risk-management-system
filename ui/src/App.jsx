import React, { useState } from 'react';
import PriceChart from './PriceChart.jsx';

export default function App() {
  const [tickers, setTickers] = useState('AAPL');
  const [data, setData] = useState({});
  const [error, setError] = useState(null);

  const fetchData = async () => {
    try {
      const params = new URLSearchParams();
      params.append('ticker', tickers);
      const response = await fetch('/stock/ticker', {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: params.toString(),
      });
      const json = await response.json();
      setData(json);
      setError(null);
    } catch (e) {
      setError('Failed to fetch data');
    }
  };

  const renderRows = () =>
    Object.entries(data).map(([ticker, prices]) => {
      const dates = Object.keys(prices);
      const latestDate = dates[dates.length - 1];
      const latestPrice = prices[latestDate];
      return (
        <tr key={ticker}>
          <td>{ticker}</td>
          <td>{latestPrice}</td>
        </tr>
      );
    });

  const firstTicker = Object.keys(data)[0];
  const firstPrices = firstTicker ? data[firstTicker] : null;
  const chartLabels = firstPrices ? Object.keys(firstPrices) : [];
  const chartData = firstPrices ? Object.values(firstPrices) : [];

  return (
    <div>
      <h1>Pension Risk Management</h1>
      <input
        type="text"
        value={tickers}
        onChange={(e) => setTickers(e.target.value)}
        placeholder="Enter comma-separated tickers"
      />
      <button onClick={fetchData}>Load</button>
      {error && <p>{error}</p>}
      <table>
        <thead>
          <tr>
            <th>Ticker</th>
            <th>Latest Close</th>
          </tr>
        </thead>
        <tbody>{renderRows()}</tbody>
      </table>
      {chartLabels.length > 0 && (
        <div style={{ maxWidth: '600px' }}>
          <PriceChart labels={chartLabels} data={chartData} />
        </div>
      )}
    </div>
  );
}
