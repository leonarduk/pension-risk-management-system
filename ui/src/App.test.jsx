import React from 'react';
import { render, screen, fireEvent, cleanup } from '@testing-library/react';
import { expect, test, vi, beforeEach, afterEach } from 'vitest';
import App from './App.jsx';

vi.mock('react-chartjs-2', () => ({
  Line: () => <div data-testid="line-chart" />,
  Scatter: () => <div data-testid="scatter-chart" />,
}));

beforeEach(() => {
  vi.restoreAllMocks();
});

afterEach(() => {
  cleanup();
});

test('loads and displays data with charts and toggles view', async () => {
  const mockTickerData = [{ Date: '2024-01-01', Close: 100 }];
  const mockRiskData = [
    { ticker: 'AAPL', annual_std: 0.1, annual_return: 0.2 },
  ];
  global.fetch = vi
    .fn()
    .mockResolvedValueOnce({ json: () => Promise.resolve(mockTickerData) })
    .mockResolvedValueOnce({ json: () => Promise.resolve(mockRiskData) });

  const { asFragment } = render(<App />);

  fireEvent.change(
    screen.getByPlaceholderText('Enter comma-separated tickers'),
    { target: { value: 'AAPL' } }
  );
  fireEvent.click(screen.getByText('Load'));

  await screen.findAllByText('AAPL');
  expect(global.fetch).toHaveBeenCalledTimes(2);
  expect(screen.getAllByText('100')[0]).toBeTruthy();
  expect(screen.getByTestId('line-chart')).toBeTruthy();
  expect(screen.getByTestId('scatter-chart')).toBeTruthy();

  fireEvent.click(screen.getByText(/View:/));
  expect(screen.getByText('1.00')).toBeTruthy();

  expect(asFragment()).toMatchSnapshot();
});

test('shows error message on fetch failure', async () => {
  global.fetch = vi.fn().mockRejectedValue(new Error('fail'));

  render(<App />);

  fireEvent.change(
    screen.getByPlaceholderText('Enter comma-separated tickers'),
    { target: { value: 'AAPL' } }
  );
  fireEvent.click(screen.getByText('Load'));

  await screen.findByText('Failed to fetch data');
  expect(screen.queryByTestId('line-chart')).toBeNull();
  expect(screen.queryByText('AAPL')).toBeNull();
});
