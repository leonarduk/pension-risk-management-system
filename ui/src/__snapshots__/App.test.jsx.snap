// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loads and displays data with charts and toggles view 1`] = `
<DocumentFragment>
  <div>
    <nav>
      <a
        href="#ticker-table"
      >
        Table
      </a>
      <a
        href="#price-chart"
      >
        Price Chart
      </a>
      <a
        href="#risk-return"
      >
        Risk/Return
      </a>
    </nav>
    <h1>
      Pension Risk Management
    </h1>
    <input
      placeholder="Enter comma-separated tickers"
      type="text"
      value="AAPL"
    />
    <button>
      Load
    </button>
    <button>
      View: relative
    </button>
    <table
      id="ticker-table"
    >
      <thead>
        <tr>
          <th>
            Ticker
          </th>
          <th>
            Latest Close
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            AAPL
          </td>
          <td>
            100
          </td>
        </tr>
      </tbody>
    </table>
    <table>
      <thead>
        <tr>
          <th>
            Ticker
          </th>
          <th>
            Latest Close
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            AAPL
          </td>
          <td>
            1.00
          </td>
        </tr>
      </tbody>
    </table>
    <div
      id="price-chart"
      style="max-width: 600px;"
    >
      <div
        data-testid="line-chart"
      />
    </div>
    <div
      id="risk-return"
      style="max-width: 600px; margin-top: 20px;"
    >
      <div
        data-testid="scatter-chart"
      />
    </div>
  </div>
</DocumentFragment>
`;
